UI/window-permissions.cpp

```diff
@@ -0,0 +1,101 @@
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+******************************************************************************/
+
+#include <QLabel>
+#include "window-permissions.hpp"
+#include "obs-app.hpp"
+
+OBSPermissions::OBSPermissions(QWidget *parent, MacPermissionStatus capture,
+			       MacPermissionStatus video,
+			       MacPermissionStatus audio,
+			       MacPermissionStatus accessibility)
+	: QDialog(parent), ui(new Ui::OBSPermissions)
+{
+	ui->setupUi(this);
+	SetStatus(ui->capturePermissionButton, capture,
+		  QTStr("MacPermissions.Item.ScreenRecording"));
+	SetStatus(ui->videoPermissionButton, video,
+		  QTStr("MacPermissions.Item.Camera"));
+	SetStatus(ui->audioPermissionButton, audio,
+		  QTStr("MacPermissions.Item.Microphone"));
+	SetStatus(ui->accessibilityPermissionButton, accessibility,
+		  QTStr("MacPermissions.Item.Accessibility"));
+}
+
+void OBSPermissions::SetStatus(QPushButton *btn, MacPermissionStatus status,
+			       const QString &preference)
+{
+	if (status == kPermissionAuthorized) {
+		btn->setText(QTStr("MacPermissions.AccessGranted"));
+	} else if (status == kPermissionNotDetermined) {
+		btn->setText(QTStr("MacPermissions.RequestAccess"));
+	} else {
+		btn->setText(
+			QTStr("MacPermissions.OpenPreferences").arg(preference));
+	}
+	btn->setEnabled(status != kPermissionAuthorized);
+	btn->setProperty("status", status);
+}
+
+void OBSPermissions::on_capturePermissionButton_clicked()
+{
+	OpenMacOSPrivacyPreferences("ScreenCapture");
+	RequestPermission(kScreenCapture);
+}
+
+void OBSPermissions::on_videoPermissionButton_clicked()
+{
+	MacPermissionStatus status =
+		(MacPermissionStatus)ui->videoPermissionButton
+			->property("status")
+			.toInt();
+	if (status == kPermissionNotDetermined) {
+		status = RequestPermission(kVideoDeviceAccess);
+		SetStatus(ui->videoPermissionButton, status,
+			  QTStr("MacPermissions.Item.Camera"));
+	} else {
+		OpenMacOSPrivacyPreferences("Camera");
+	}
+}
+
+void OBSPermissions::on_audioPermissionButton_clicked()
+{
+	MacPermissionStatus status =
+		(MacPermissionStatus)ui->audioPermissionButton
+			->property("status")
+			.toInt();
+	if (status == kPermissionNotDetermined) {
+		status = RequestPermission(kAudioDeviceAccess);
+		SetStatus(ui->audioPermissionButton, status,
+			  QTStr("MacPermissions.Item.Microphone"));
+	} else {
+		OpenMacOSPrivacyPreferences("Microphone");
+	}
+}
+
+void OBSPermissions::on_accessibilityPermissionButton_clicked()
+{
+	OpenMacOSPrivacyPreferences("Accessibility");
+	RequestPermission(kAccessibility);
+}
+
+void OBSPermissions::on_continueButton_clicked()
+{
+	config_set_bool(GetGlobalConfig(), "General",
+			"MacOSPermissionDialogShown", true);
```

Comment:
Would it be worth versioning the value in case Apple/we add new permissions in future so that we can forcefully reopen the dialog when requirements change? Although it may be better to just use the notification PR and display a message that way when things change in future.

I just don't want to potentially limit ourselves this time.

Reply:
Good idea.

Reply:
Why not just pop the dialog if any permission is not given? Maybe guard if with the `DEBUG` macro, because it would pop open on _any_ code change during development otherwise.

Reply:
I personally wouldn't want the dialog to pop up each time in my dev environment, there are many things that don't require any permissions and having to click "Continue" each time you open OBS would be an annoyance.

Reply:
Added versioning.